File: String/stringHashing.py
text = input()

n = int(input())

h = [0] * (len(text))
p = [0] * (len(text))

h[0] = ord(text[0])
p[0] = 1


A = 911382323
B = 2**63 - 1

for i in range(1,len(text)):    
    h[i] = (h[i-1]*A + ord(text[i])) % B
    p[i] = (p[i-1]*A) % B
    


for i in range(n):
    a,b = map(int, input().split())
    b -= 1
    if a == 0:
        print(h[b])
    else:
        print((h[b]-h[a-1]*p[b-a+1]) % B)
        
    
 

File: String/stringHashing.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

func main() {

	scanner := bufio.NewScanner(os.Stdin)

	var text string
	scanner.Scan()
	text = scanner.Text()
	println(len(text))
	scanner.Scan()
	n := parse(scanner.Text())

	prime := (1 << 19) - 1
	a := 26
	dict := make(map[string]int)
	println(n)
	for i := 0; i < n; i++ {
		scanner.Scan()
		line := scanner.Text()
		var L, R int
		fmt.Sscanf(line, "%d %d", &L, &R)
		substring := text[L:R]
		hashVal := 0

		if dict[substring] == 0 {
			for _, ch := range substring {
				hashVal = (hashVal*a + int(ch)) % prime
			}
			dict[substring] = hashVal
		} else {
			hashVal = dict[substring]
		}

		fmt.Println(hashVal)

	}
}

func parse(s string) int {
	var n int
	fmt.Sscanf(s, "%d", &n)
	return n
}
